Running on a local cluster

I'd like to have parsl push jobs out to a small cluster we've got in the HEP group. About 120 cores, split over about 10 machines. There is no batch manager or similar.

The pilot model seemed appropriate. I was thinking to use the IPyParallelExecutor, and then have each on a seperate note.

However, I don't know what to use for a provider. LocalProvider seems like what I want - tho it needs to be on another machine. I tried to hack it by giving it an SSHChannel, but that didn't work.

Here is the configuration I've been playing with:

        remote_config = Config(
            executors=[
                IPyParallelExecutor(
                    label='remote_ipp',
                    provider=LocalProvider(
                        min_blocks=1,
                        init_blocks=1,
                        max_blocks=4,
                        tasks_per_node=2,
                        nodes_per_block=1,
                        parallelism=0.5,
                        channel=SSHChannel(hostname=m)
                    )
                ) for m in args.remote
            ]
        )
        parsl.load(remote_config)

If I remove the "channel=" line, btw, then my code runs just fine. All the jobs seem to execute without difficulty. Though the IPy executor generates a huge number of log messages!

I'm a totally newbee here - This is the first time I've used parsl. Very nice, btw, in about 2 hours I had something up and running on a single machine. Took me another 2 days to make it "robust". :-)

Ok - crash, again, because it couldn't find the parsl module:

```
(parsl) -bash-4.2$ time python ../../DistributedRunner/build/run_distributed.py test/tracks 4500-4501 junkdir tracks.root --cmd_prefix "source $ATLAS_LOCAL_ROOT_BASE/user/atlasLocalSetup.sh" --cmd_prefix "lsetup \"root recommended\"" --cmd_prefix "export TESTSTANDCONFIG=/phys/groups/tev/scratch3/users/gwatts/MATHUSLA/TestStandOffline/config" --cmd_prefix 'export LD_LIBRARY_PATH=/phys/groups/tev/scratch3/users/gwatts/MATHUSLA/TestStandOffline/build:$LD_LIBRARY_PATH' --remote tev02
/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/ipython-7.2.0-py3.7.egg/IPython/paths.py:48: UserWarning: Ignoring ~/.config/ipython in favour of ~/.ipython. Remove ~/.config/ipython to get rid of this message
  'get rid of this message').format(cu(xdg_ipdir), cu(ipdir)))
Traceback (most recent call last):
  File "../../DistributedRunner/build/run_distributed.py", line 287, in <module>
    run_list = get_run_list(args.run_list, args.cmd_prefix)
  File "../../DistributedRunner/build/run_distributed.py", line 216, in get_run_list
    v = run_bash_command("%s echo NOWAYDUDE && %s %s" % (pCmd, exe_dir, runinfo), stdout='get_run_list.log').result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/parsl-0.6.2a1-py3.7.egg/parsl/dataflow/futures.py", line 126, in parent_callback
    res = executor_fu.result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/parsl-0.6.2a1-py3.7.egg/parsl/dataflow/dflow.py", line 257, in handle_exec_update
    res = future.result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/ipyparallel-6.2.3-py3.7.egg/ipyparallel/client/asyncresult.py", line 226, in _resolve_result
    raise r
ipyparallel.error.RemoteError: ModuleNotFoundError(No module named 'parsl')
```

poking around (sorry, not 100% srue what I'm looking for), I see that there is a lot file in runinfo/000 and found parsl.log, which I think is the most useful?

```
(parsl) -bash-4.2$ cat parsl.log
2018-12-04 16:25:17 parsl.dataflow.dflow:82 [DEBUG]  Starting DataFlowKernel with config
Config(
    app_cache=True,
    checkpoint_files=None,
    checkpoint_mode=None,
    checkpoint_period=None,
    data_management_max_threads=10,
    executors=[IPyParallelExecutor(
        container_image=None,
        controller=Controller(
            interfaces=None,
            ipython_dir='~/.ipython',
            log=True,
            mode='auto',
            port=None,
            port_range=None,
            profile='default',
            public_ip=None,
            reuse=False
        ),
        engine_debug_level=None,
        engine_dir=None,
        label='remote_ipp_tev02',
        managed=True,
        provider=LocalProvider(
            channel=SSHChannel(
                'tev02',
                envs={},
                password=None,
                script_dir=None,
                username=None
            ),
            cmd_timeout=30,
            init_blocks=1,
            launcher=SingleNodeLauncher(),
            max_blocks=4,
            min_blocks=1,
            nodes_per_block=1,
            parallelism=0.5,
            walltime='00:15:00',
            worker_init='conda activate parsl'
        ),
        storage_access=[],
        workers_per_node=10,
        working_dir=None
    )],
    lazy_errors=True,
    monitoring_config=None,
    retries=0,
    run_dir='runinfo',
    strategy='simple',
    usage_tracking=True
)
2018-12-04 16:25:17 parsl.dataflow.dflow:83 [INFO]  Parsl version: 0.6.2-a1
2018-12-04 16:25:17 parsl.dataflow.usage_tracking.usage:126 [DEBUG]  Tracking status: True
2018-12-04 16:25:17 parsl.dataflow.usage_tracking.usage:127 [DEBUG]  Testing mode   : False
2018-12-04 16:25:17 parsl.dataflow.dflow:123 [INFO]  Run id is: 27937f6e-1e36-45f2-8480-c1a7905de845
2018-12-04 16:25:17 parsl.dataflow.memoization:52 [INFO]  App caching initialized
2018-12-04 16:25:19 parsl.executors.ipp_controller:83 [DEBUG]  Starting ipcontroller with 'ipcontroller --ipython-dir=/phys/groups/tev/scratch3/users/gwatts/MATHUSLA/TestStandOffline/build/runinfo/000 --ip=* --profile=remote_ipp_tev02   '
2018-12-04 16:25:21 parsl.executors.ipp:107 [DEBUG]  Waiting for /phys/groups/tev/scratch3/users/gwatts/MATHUSLA/TestStandOffline/build/runinfo/000/profile_remote_ipp_tev02/security/ipcontroller-client.json
2018-12-04 16:25:21 parsl.executors.ipp:123 [DEBUG]  Starting IPyParallelExecutor with provider:
LocalProvider(
    channel=SSHChannel(
        'tev02',
        envs={},
        password=None,
        script_dir='/phys/groups/tev/scratch3/users/gwatts/MATHUSLA/TestStandOffline/build/runinfo/000/submit_scripts',
        username=None
    ),
    cmd_timeout=30,
    init_blocks=1,
    launcher=SingleNodeLauncher(),
    max_blocks=4,
    min_blocks=1,
    nodes_per_block=1,
    parallelism=0.5,
    walltime='00:15:00',
    worker_init='conda activate parsl'
)
2018-12-04 16:25:21 parsl.providers.local.local:188 [DEBUG]  Not a localChannel, files need to be moved
2018-12-04 16:25:24 parsl.executors.ipp:242 [DEBUG]  Launched block 0:8346
2018-12-04 16:25:24 parsl.executors.ipp:249 [ERROR]  No execution provider available
2018-12-04 16:25:24 parsl.executors.ipp:132 [DEBUG]  Starting executor
2018-12-04 16:25:24 parsl.dataflow.strategy:125 [DEBUG]  Scaling strategy: simple
2018-12-04 16:25:24 parsl.dataflow.dflow:669 [INFO]  Task 0 submitted for App remote_side_bash_executor, waiting on tasks []
2018-12-04 16:25:24 parsl.dataflow.dflow:679 [DEBUG]  Task 0 set to pending state with AppFuture: <AppFuture at 0x7f1f0f22b4a8 state=pending>
2018-12-04 16:25:24 parsl.dataflow.dflow:472 [INFO]  Task 0 launched on executor remote_ipp_tev02
2018-12-04 16:25:27 parsl.dataflow.strategy:202 [DEBUG]  Executor remote_ipp_tev02 has 1 active tasks, 0/0/0 running/submitted/pending blocks, and 0 connected engines
2018-12-04 16:25:27 parsl.dataflow.strategy:256 [DEBUG]  Requesting 1 more blocks
2018-12-04 16:25:27 parsl.providers.local.local:188 [DEBUG]  Not a localChannel, files need to be moved
2018-12-04 16:25:29 parsl.dataflow.dflow:262 [ERROR]  Task 0 failed
Traceback (most recent call last):
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/parsl-0.6.2a1-py3.7.egg/parsl/dataflow/dflow.py", line 257, in handle_exec_update
    res = future.result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/phys/groups/tev/scratch3/users/gwatts/anaconda3/envs/parsl/lib/python3.7/site-packages/ipyparallel-6.2.3-py3.7.egg/ipyparallel/client/asyncresult.py", line 226, in _resolve_result
    raise r
ipyparallel.error.RemoteError: ModuleNotFoundError(No module named 'parsl')
2018-12-04 16:25:29 parsl.dataflow.dflow:286 [INFO]  Task 0 failed after 0 retry attempts
2018-12-04 16:25:29 parsl.dataflow.dflow:784 [INFO]  DFK cleanup initiated
2018-12-04 16:25:29 parsl.dataflow.dflow:715 [INFO]  Summary of tasks in DFK:
2018-12-04 16:25:29 parsl.dataflow.dflow:746 [INFO]  Tasks in state States.failed: 0
2018-12-04 16:25:29 parsl.dataflow.dflow:753 [INFO]  End of summary
2018-12-04 16:25:29 parsl.dataflow.dflow:808 [INFO]  Terminating flow_control and strategy threads
2018-12-04 16:25:30 parsl.executors.ipp:242 [DEBUG]  Launched block 0:8636
2018-12-04 16:25:30 parsl.executors.ipp:249 [ERROR]  No execution provider available
2018-12-04 16:25:33 parsl.executors.ipp:295 [DEBUG]  IPP:Shutdown sequence: Attempting controller kill
2018-12-04 16:25:33 parsl.executors.ipp_controller:143 [DEBUG]  Controller exited with 0
2018-12-04 16:25:33 parsl.executors.ipp:305 [DEBUG]  Done with executor shutdown
2018-12-04 16:25:33 parsl.data_provider.data_manager:98 [DEBUG]  Done with executor shutdown
2018-12-04 16:25:33 parsl.dataflow.dflow:831 [INFO]  DFK cleanup complete
```

When I watch tev02 with `top` it looks like the `conda` command does actually execute. Perhaps the SSH channel isn't persistent? Or it isn't executed early enough, or it is executed as a sub-process, but the parent process isn't getting the benifit?

If there are other log files you'd like me to upload, let me know!

--- a/parsl/providers/local/local.py
+++ b/parsl/providers/local/local.py
@@ -224,15 +224,16 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
         for job in job_ids:
             logger.debug("Terminating job/proc_id : {0}".format(job))
             # Here we are assuming that for local, the job_ids are the process id's
-            if self.resouces[job]['proc'] :
+            if self.resources[job]['proc'] :
                 proc = self.resources[job]['proc']
                 os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
                 self.resources[job]['status'] = 'CANCELLED'
-            elif self.resource[job]['remote_pid']:
-                cmd = "kill -- -$(ps -o pgid={} | grep -o '[0-9]*')".format(self.resource[job]['remote_pid'])
+
+            elif self.resources[job]['remote_pid']:
+                cmd = "kill -- -$(ps -o pgid={} | grep -o '[0-9]*')".format(self.resources[job]['remote_pid'])
                 retcode, stdout, stderr = self.channel.execute_wait(cmd, self.cmd_timeout)
                 if retcode != 0:
-                    logger.warning("Failed to kill PID:{} and child processes on {}".format(self.resource[job]['remote_pid'],
+                    logger.warning("Failed to kill PID:{} and child processes on {}".format(self.resources[job]['remote_pid'],
                                                                                             self.label))
 
         rets = [True for i in job_ids]

--- a/parsl/providers/local/local.py
+++ b/parsl/providers/local/local.py
@@ -52,12 +52,14 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
                  min_blocks=0,
                  max_blocks=10,
                  walltime="00:15:00",
+                 worker_init='',
                  parallelism=1):
         self.channel = channel
         self.label = 'local'
         self.provisioned_blocks = 0
         self.nodes_per_block = nodes_per_block
         self.launcher = launcher
+        self.worker_init = worker_init
         self.init_blocks = init_blocks
         self.min_blocks = min_blocks
         self.max_blocks = max_blocks
@@ -81,16 +83,30 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
 
         logging.debug("Checking status of : {0}".format(job_ids))
         for job_id in self.resources:
-            poll_code = self.resources[job_id]['proc'].poll()
-            if self.resources[job_id]['status'] in ['COMPLETED', 'FAILED']:
-                continue
 
-            if poll_code is None:
-                self.resources[job_id]['status'] = 'RUNNING'
-            elif poll_code == 0 and self.resources[job_id]['status'] != 'RUNNING':
-                self.resources[job_id]['status'] = 'COMPLETED'
-            elif poll_code < 0 and self.resources[job_id]['status'] != 'RUNNING':
-                self.resources[job_id]['status'] = 'FAILED'
+            if self.resources[job_id]['proc']:
+
+                poll_code = self.resources[job_id]['proc'].poll()
+                if self.resources[job_id]['status'] in ['COMPLETED', 'FAILED']:
+                    continue
+
+                if poll_code is None:
+                    self.resources[job_id]['status'] = 'RUNNING'
+                elif poll_code == 0 and self.resources[job_id]['status'] != 'RUNNING':
+                    self.resources[job_id]['status'] = 'COMPLETED'
+                elif poll_code < 0 and self.resources[job_id]['status'] != 'RUNNING':
+                    self.resources[job_id]['status'] = 'FAILED'
+
+            elif self.resources[job_id]['remote_pid']:
+
+                retcode, stdout, stderr = self.channel.execute_wait('ps -p {} &> /dev/null; echo "STATUS:$?" ', 3)
+                for line in stdout.split('\n'):
+                    if line.startswith("STATUS:"):
+                        status = line.split("STATUS:")[1].strip()
+                        if status == "0":
+                            self.resources[job_id]['status'] = 'RUNNING'
+                        else:
+                            self.resources[job_id]['status'] = 'FAILED'
 
         return [self.resources[jid]['status'] for jid in job_ids]
 
@@ -158,12 +174,35 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
         script_path = "{0}/{1}.sh".format(self.script_dir, job_name)
         script_path = os.path.abspath(script_path)
 
-        wrap_command = self.launcher(command, tasks_per_node, self.nodes_per_block)
+        wrap_command = self.worker_init + '\n' + self.launcher(command, tasks_per_node, self.nodes_per_block)
 
         self._write_submit_script(wrap_command, script_path)
 
-        job_id, proc = self.channel.execute_no_wait('bash {0}'.format(script_path), 3)
-        self.resources[job_id] = {'job_id': job_id, 'status': 'RUNNING', 'blocksize': blocksize, 'proc': proc}
+        proc = None
+        remote_pid = None
+        if not isinstance(self.channel, LocalChannel):
+            logger.debug("Not a localChannel, files need to be moved")
+            script_path = self.channel.push_file(script_path, self.channel.script_dir)
+            cmd = 'bash {0} & \necho "PID:$!" '.format(script_path)
+            retcode, stdout, stderr = self.channel.execute_wait(cmd, 3)
+            logger.debug("Stdout : {}".format(stdout))
+            logger.debug("Stderr : {}".format(stderr))
+            for line in stdout.split('\n'):
+                if line.startswith("PID:"):
+                    remote_pid = line.split("PID:")[1].strip()
+                    job_id = remote_pid
+        else:
+
+            try:
+                job_id, proc = self.channel.execute_no_wait('bash {0}'.format(script_path), 3)
+            except Exception as e:
+                logger.debug("Channel execute failed for:{}, {}".format(self.channel, e))
+                raise
+
+        self.resources[job_id] = {'job_id': job_id, 'status': 'RUNNING',
+                                  'blocksize': blocksize,
+                                  'remote_pid': remote_pid,
+                                  'proc': proc}
 
         return job_id
 

--- a/parsl/providers/local/local.py
+++ b/parsl/providers/local/local.py
@@ -224,7 +224,7 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
         for job in job_ids:
             logger.debug("Terminating job/proc_id : {0}".format(job))
             # Here we are assuming that for local, the job_ids are the process id's
-            if self.resources[job]['proc'] :
+            if self.resources[job]['proc']:
                 proc = self.resources[job]['proc']
                 os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
                 self.resources[job]['status'] = 'CANCELLED'

--- a/parsl/providers/local/local.py
+++ b/parsl/providers/local/local.py
@@ -53,6 +53,7 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
                  max_blocks=10,
                  walltime="00:15:00",
                  worker_init='',
+                 cmd_timeout=30,
                  parallelism=1):
         self.channel = channel
         self.label = 'local'
@@ -66,6 +67,7 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
         self.parallelism = parallelism
         self.walltime = walltime
         self.script_dir = None
+        self.cmd_timeout = cmd_timeout
 
         # Dictionary that keeps track of jobs, keyed on job_id
         self.resources = {}
@@ -99,7 +101,8 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
 
             elif self.resources[job_id]['remote_pid']:
 
-                retcode, stdout, stderr = self.channel.execute_wait('ps -p {} &> /dev/null; echo "STATUS:$?" ', 3)
+                retcode, stdout, stderr = self.channel.execute_wait('ps -p {} &> /dev/null; echo "STATUS:$?" ',
+                                                                    self.cmd_timeout)
                 for line in stdout.split('\n'):
                     if line.startswith("STATUS:"):
                         status = line.split("STATUS:")[1].strip()
@@ -178,23 +181,26 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
 
         self._write_submit_script(wrap_command, script_path)
 
+        job_id = None
         proc = None
         remote_pid = None
         if not isinstance(self.channel, LocalChannel):
             logger.debug("Not a localChannel, files need to be moved")
             script_path = self.channel.push_file(script_path, self.channel.script_dir)
-            cmd = 'bash {0} & \necho "PID:$!" '.format(script_path)
-            retcode, stdout, stderr = self.channel.execute_wait(cmd, 3)
-            logger.debug("Stdout : {}".format(stdout))
-            logger.debug("Stderr : {}".format(stderr))
+
+            # Bash would return until the streams are closed. So we redirect to a outs file
+            cmd = 'bash {0} &> {0}.out & \n echo "PID:$!" '.format(script_path)
+            retcode, stdout, stderr = self.channel.execute_wait(cmd, self.cmd_timeout)
             for line in stdout.split('\n'):
                 if line.startswith("PID:"):
                     remote_pid = line.split("PID:")[1].strip()
                     job_id = remote_pid
+            if job_id is None:
+                logger.warning("Channel failed to start remote command/retrieve PID")
         else:
 
             try:
-                job_id, proc = self.channel.execute_no_wait('bash {0}'.format(script_path), 3)
+                job_id, proc = self.channel.execute_no_wait('bash {0}'.format(script_path), self.cmd_timeout)
             except Exception as e:
                 logger.debug("Channel execute failed for:{}, {}".format(self.channel, e))
                 raise
@@ -218,11 +224,18 @@ class LocalProvider(ExecutionProvider, RepresentationMixin):
         for job in job_ids:
             logger.debug("Terminating job/proc_id : {0}".format(job))
             # Here we are assuming that for local, the job_ids are the process id's
-            proc = self.resources[job]['proc']
-            os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
-            self.resources[job]['status'] = 'CANCELLED'
-        rets = [True for i in job_ids]
+            if self.resouces[job]['proc'] :
+                proc = self.resources[job]['proc']
+                os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
+                self.resources[job]['status'] = 'CANCELLED'
+            elif self.resource[job]['remote_pid']:
+                cmd = "kill -- -$(ps -o pgid={} | grep -o '[0-9]*')".format(self.resource[job]['remote_pid'])
+                retcode, stdout, stderr = self.channel.execute_wait(cmd, self.cmd_timeout)
+                if retcode != 0:
+                    logger.warning("Failed to kill PID:{} and child processes on {}".format(self.resource[job]['remote_pid'],
+                                                                                            self.label))
 
+        rets = [True for i in job_ids]
         return rets
 
     @property

--- a/parsl/tests/configs/htex_ad_hoc_cluster.py
+++ b/parsl/tests/configs/htex_ad_hoc_cluster.py
@@ -1,6 +1,5 @@
 from parsl.providers import LocalProvider
 from parsl.channels import SSHChannel
-from parsl.launchers import SingleNodeLauncher
 from parsl.executors import HighThroughputExecutor
 
 from parsl.config import Config

--- a/parsl/tests/configs/htex_ad_hoc_cluster.py
+++ b/parsl/tests/configs/htex_ad_hoc_cluster.py
@@ -1,0 +1,30 @@
+from parsl.providers import LocalProvider
+from parsl.channels import SSHChannel
+from parsl.launchers import SingleNodeLauncher
+from parsl.executors import HighThroughputExecutor
+
+from parsl.config import Config
+
+username = "yadunand"
+remotes = ['midway2-login1.rcc.uchicago.edu', 'midway2-login2.rcc.uchicago.edu']
+
+config = Config(
+    executors=[
+        HighThroughputExecutor(
+            label='remote_ipp_{}'.format(m),
+            cores_per_worker=4,
+            worker_debug=False,
+            address="128.135.112.73",
+            provider=LocalProvider(
+                init_blocks=1,
+                nodes_per_block=1,
+                parallelism=0.5,
+                worker_init="source /scratch/midway2/yadunand/parsl_env_setup.sh",
+                channel=SSHChannel(hostname=m,
+                                   username=username,
+                                   script_dir="/scratch/midway2/{}/parsl_tests/".format(username)
+                )
+            )
+        ) for m in remotes
+    ],
+)

--- a/parsl/tests/configs/htex_ad_hoc_cluster.py
+++ b/parsl/tests/configs/htex_ad_hoc_cluster.py
@@ -10,7 +10,7 @@ remotes = ['midway2-login1.rcc.uchicago.edu', 'midway2-login2.rcc.uchicago.edu']
 config = Config(
     executors=[
         HighThroughputExecutor(
-            label='remote_ipp_{}'.format(m),
+            label='remote_htex_{}'.format(m),
             cores_per_worker=4,
             worker_debug=False,
             address="128.135.112.73",

--- a/parsl/tests/configs/ipp_ad_hoc_cluster.py
+++ b/parsl/tests/configs/ipp_ad_hoc_cluster.py
@@ -1,6 +1,5 @@
 from parsl.providers import LocalProvider
 from parsl.channels import SSHChannel
-from parsl.launchers import SingleNodeLauncher
 from parsl.executors.ipp_controller import Controller
 from parsl.executors.ipp import IPyParallelExecutor
 

--- a/parsl/tests/configs/ipp_ad_hoc_cluster.py
+++ b/parsl/tests/configs/ipp_ad_hoc_cluster.py
@@ -1,0 +1,31 @@
+from parsl.providers import LocalProvider
+from parsl.channels import SSHChannel
+from parsl.launchers import SingleNodeLauncher
+from parsl.executors.ipp_controller import Controller
+from parsl.executors.ipp import IPyParallelExecutor
+
+from parsl.config import Config
+
+username = "yadunand"
+remotes = ['midway2-login1.rcc.uchicago.edu', 'midway2-login2.rcc.uchicago.edu']
+
+config = Config(
+    executors=[
+        IPyParallelExecutor(
+            label='remote_ipp_{}'.format(m),
+            workers_per_node=2,  # Replaces provider.tasks_per_node
+            engine_debug_level="DEBUG",
+            controller=Controller(public_ip="128.135.112.73"),
+            provider=LocalProvider(
+                init_blocks=1,
+                nodes_per_block=1,
+                parallelism=0.5,
+                worker_init="source /scratch/midway2/yadunand/parsl_env_setup.sh",
+                channel=SSHChannel(hostname=m,
+                                   username=username,
+                                   script_dir="/scratch/midway2/{}/parsl_tests/".format(username)
+                )
+            ),
+        ) for m in remotes
+    ],
+)

